File: generated_code/GeminiPro/Java/047_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-balancing binary search tree with the additional property that
 * recently accessed elements are quick to access again. It performs basic operations
 * such as insertion, deletion, and search. After any of these operations on a key, the
 * node containing that key (or the last node accessed in a search) is moved to the
 * root of the tree through a series of rotations.
 */
public class SplayTree {

    private Node root;

    /**
     * Inner class to represent a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Private Helper Methods ---

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. Brings the node with the given key to the root of the
     * subtree. If the key is not found, the last accessed node is brought to the root.
     * This is a recursive, bottom-up splay implementation.
     *
     * @param node The root of the current subtree.
     * @param key The key to search for and splay.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node node, int key) {
        // Base cases: node is null or key is present at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key lies in left subtree
        if (key < node.key) {
            // Key is not in tree, we are done
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left Left)
            if (key < node.left.key) {
                node.left.left = splay(node.left.left, key);
                node = rightRotate(node);
            }
            // Zig-Zag (Left Right)
            else if (key > node.left.key) {
                node.left.right = splay(node.left.right, key);
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }

            // Do second rotation for node (if required)
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key lies in right subtree
        else {
            // Key is not in tree, we are done
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right Left)
            if (key < node.right.key) {
                node.right.left = splay(node.right.left, key);
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > node.right.key) {
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            }

            // Do second rotation for node (if required)
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    // --- Public API Methods ---

    /**
     * Searches for a key in the tree.
     * If the key is found, it is splayed to the root. If not found, the last
     * accessed node on the search path is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a new key into the Splay Tree.
     * If the key already exists, the tree is splayed on that key, and no new node is inserted.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // Case 1: The tree is empty.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // Case 2: The key is already in the tree. Do nothing.
        if (root.key == key) {
            return;
        }

        // Case 3: Insert the new key.
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key. If the key exists, it becomes the root.
        root = splay(root, key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root.
        // We need to join its left and right subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Find the maximum element in the left subtree by traversing to the rightmost node.
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // Splay the left subtree around its maximum element.
            // This makes the maximum element the root of the modified left subtree.
            leftSubtree = splay(leftSubtree, maxNode.key);

            // The new root of the left subtree now has no right child.
            // Attach the original right subtree as its right child.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/047_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Python class that implements a dictionary-like set for integers
    using a Splay Tree data structure.

    The tree automatically rebalances itself by moving frequently accessed elements
    closer to the root through a process called "splaying".

    Methods:
    - insert(key): Adds an integer key to the set.
    - delete(key): Removes an integer key from the set.
    - search(key): Checks for the existence of a key and splays the accessed node.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left is not None:
            y.left.parent = x

        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right is not None:
            y.right.parent = x

        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent is not None:
            p = x.parent
            g = p.parent
            if g is None:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-zig case (left-left)
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-zig case (right-right)
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-zag case (left-right)
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-zag case (right-left)
                self._right_rotate(p)
                self._left_rotate(g)

    def _maximum(self, node):
        """Finds the node with the maximum key in a subtree."""
        while node.right is not None:
            node = node.right
        return node

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last accessed node (the would-be parent)
        is splayed to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        last_node = None
        current = self.root
        while current is not None:
            last_node = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                self._splay(current)
                return True

        if last_node is not None:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        p = None
        current = self.root
        while current is not None:
            p = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and we are done.
                self._splay(current)
                return

        # Key is not in the tree, insert a new node
        new_node = self._Node(key)
        new_node.parent = p

        if p is None:  # The tree was empty
            self.root = new_node
        elif key < p.key:
            p.left = new_node
        else:
            p.right = new_node

        # Splay the newly inserted node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The standard splay tree deletion algorithm is used:
        1. The node to be deleted is splayed to the root.
        2. The root is removed, splitting the tree into two subtrees.
        3. The maximum element of the left subtree is splayed to its root.
        4. The right subtree is attached as the right child of the new root.

        If the key is not found, the would-be parent is splayed as a side effect
        of the initial search.

        Args:
            key (int): The integer key to delete.
        """
        # Search for the key. This splays the node (or its parent) to the root.
        if not self.search(key):
            # Key was not in the tree. search() already splayed the parent.
            return

        # At this point, the node to delete is at the root because search(key)
        # returned True and splayed it.
        z = self.root
        left_subtree = z.left
        right_subtree = z.right

        if left_subtree is None and right_subtree is None:
            self.root = None
        elif left_subtree is None:
            self.root = right_subtree
            self.root.parent = None
        elif right_subtree is None:
            self.root = left_subtree
            self.root.parent = None
        else:
            # Both subtrees exist.
            # Detach the left subtree from the old root.
            left_subtree.parent = None

            # Find the maximum node in the left subtree.
            max_in_left = self._maximum(left_subtree)

            # Splay this max node to the root of the (now separated) left subtree.
            # We can do this by temporarily setting the tree's root to the left
            # subtree's root and then using the existing search/splay logic.
            self.root = left_subtree
            self.search(max_in_left.key)  # This splays max_in_left

            # After the splay, self.root is max_in_left, which has no right child.
            # Attach the original right subtree.
            self.root.right = right_subtree
            right_subtree.parent = self.root